### PR TITLE
Fix select from system.dictionaries when there is dictionary with bad structure

### PR DESCRIPTION
**Pull request by aalexfvk**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `SELECT ... FROM system.dictionaries` exception when there is a dictionary with a bad structure (e.g. incorrect type in xml config)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/System/StorageSystemDictionaries.cpp [hunk 1]
```diff
@@ -20,6 +20,72 @@
...
+        LoadingStartTime,
+        LastSuccessfulUpdateTime,
+        LoadingDuration,
+        LastException,
+        Comment,
+        /// Virtual columns
+        Key
+    };
+
+    std::shared_ptr<const IDictionary> getDictionaryPtr(const ExternalLoader::LoadResult & load_result)
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemDictionaries.cpp:57]: Consider to remove this method, it just implements `dynamic_pointer_cast`.

File: src/Storages/System/StorageSystemDictionaries.cpp [hunk 2]
```diff
@@ -20,6 +20,72 @@
 
 namespace DB
 {
+namespace
+{
+    /// Used for column indexes!  Don't change the numbers!
+    enum SystemDictionariesColumn
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemDictionaries.cpp:26]: Need to fix indentation. In anonymous namespace there are no indentation.

File: src/Storages/System/StorageSystemDictionaries.cpp [hunk 3]
```diff
@@ -70,96 +135,84 @@ void StorageSystemDictionaries::fillData(MutableColumns & res_columns, ContextPt
     if (!check_access_for_dictionaries)
         return;
 
-    for (const auto & load_result : external_dictionaries.getLoadResults())
+    for (size_t index = 0; const auto & load_result : external_dictionaries.getLoadResults())
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemDictionaries.cpp:138]: `external_dictionaries.getLoadResults()` returns vector. We can just write index loop over it. Without introducing additional `index` variable.

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+    <clickhouse>
```
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml:1]: Need to fix indentation.

File: src/Storages/System/StorageSystemDictionaries.cpp [hunk 5]
```diff
@@ -70,96 +135,84 @@ void StorageSystemDictionaries::fillData(MutableColumns & res_columns, ContextPt
...
+            static_cast<UInt64>(std::chrono::system_clock::to_time_t(load_result.last_successful_update_time)));
+        res_columns[SystemDictionariesColumn::LoadingDuration]->insert(
+            std::chrono::duration_cast<std::chrono::duration<float>>(load_result.loading_duration).count());
 
-        res_columns[i++]->insert(dictionary_structure.getKeyDescription());
+        if (last_exception)
+            res_columns[SystemDictionariesColumn::LastException]->insert(getExceptionMessage(last_exception, false));
+
+        index++;
+        /// Set to default unfilled columns
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemDictionaries.cpp:211]: Better to rename something like `Add default value to unfilled columns`.

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml (hunk 4 above)
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml:1]: It make sense to rename this file into `invalid_dict`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Comment by kitaisreal** [comment]
@aalexfvk there is some issue with CLA assistant that need to be fixed.

File: src/Storages/System/StorageSystemDictionaries.cpp (hunk 1 above)
Comment (aalexfvk) [reply src/Storages/System/StorageSystemDictionaries.cpp]: Removed

File: src/Storages/System/StorageSystemDictionaries.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aalexfvk) [reply src/Storages/System/StorageSystemDictionaries.cpp]: I agree. Removed it and made more consistent

Comment (aalexfvk) [reply src/Storages/System/StorageSystemDictionaries.cpp]: Fixed it, although it was done by the auto-formatter according to [this](https://github.com/ClickHouse/ClickHouse/blob/245899c0f66e4f3e5d3d833dc3cd2ffc5cbd6cb6/.clang-format#L82). BTW in the project in different places differently

File: src/Storages/System/StorageSystemDictionaries.cpp (hunk 3 above)
Comment (aalexfvk) [reply src/Storages/System/StorageSystemDictionaries.cpp]: Changed the code. No longer relevant

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml (hunk 4 above)
Comment (aalexfvk) [reply tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml]: Fixed

File: src/Storages/System/StorageSystemDictionaries.cpp (hunk 5 above)
Comment (aalexfvk) [reply src/Storages/System/StorageSystemDictionaries.cpp]: Changed the code. No longer relevant

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml (hunk 4 above)
Comment (aalexfvk) [reply tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/bad_dict.xml]: Done

File: tests/integration/test_dictionaries_with_bad_structure/test.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+DICTIONARY_FILES = [
+    "configs/dictionaries/invalid_dict.xml",
+    "configs/dictionaries/valid_dict.xml",
+]
+
+cluster = ClickHouseCluster(__file__)
+instance = cluster.add_instance("instance", dictionaries=DICTIONARY_FILES)
+
+VALID_DICT_NAME = "valid_dict"
+BAD_DICT_NAME = "invalid_dict"
```
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/test.py:15]: Consider to rename `BAD_DICT_NAME` into `INVALID_DICT_NAME`

File: tests/integration/test_dictionaries_with_bad_structure/test.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def test_select_from_system_dictionaries_with_bad_dictionary(started_cluster):
```
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/test.py:28]: Consider to rename `test_select_from_system_dictionaries_with_bad_dictionary` with `test_select_from_system_dictionaries_with_invalid_dictionary`.

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/invalid_dict.xml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            <cache><size_in_cells>128</size_in_cells></cache>
+        </layout>
+
+        <structure>
+            <id>
+                <name>id</name>
+            </id>
+
+            <attribute>
+                <name>bad_attr</name>
```
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/invalid_dict.xml:28]: Consider to rename into `invalid_attr`.

File: tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/invalid_dict.xml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+                <type>bad_type</type>
```
Comment (kitaisreal) [inline tests/integration/test_dictionaries_with_bad_structure/configs/dictionaries/invalid_dict.xml:29]: Consider to rename into `invalid_type`.

**Review by kitaisreal — APPROVED** [review]

**Comment by kitaisreal** [comment]
@aalexfvk maybe there are some issue with CLA for your commits, consider adding commit e-mail addresses into your github account.

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**